### ISSUE TITLE
LLVM occassionally crashes when compiling project

### ISSUE THREAD

**Issue by JakeChampion** [llvm/llvm-project#100084]
The project sometimes builds and sometimes fails to build with the below backtrace

I'm not sure what other information to give here, other than the backtrace - please let me know what (if any) other information would be useful to include and I will try my best to return with that information 🙇 

```
0:12.22 PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
 0:12.32  #0 0x000000010aba7df0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb83df0)
 0:12.32  #1 0x000000010aba835c SignalHandler(int) (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb8435c)
 0:12.32  #2 0x000000019436b584 (/usr/lib/system/libsystem_platform.dylib+0x180477584)
 0:12.32  #3 0x0000000104f7e784 lld::wasm::InputChunk::relocate(unsigned char*) const (/Users/jakechampion/.mozbuild/clang/bin/lld+0x100322784)
 0:12.32  #4 0x0000000104f7e784 lld::wasm::InputChunk::relocate(unsigned char*) const (/Users/jakechampion/.mozbuild/clang/bin/lld+0x100322784)
 0:12.32  #5 0x0000000104f97e18 lld::wasm::CustomSection::writeTo(unsigned char*) (/Users/jakechampion/.mozbuild/clang/bin/lld+0x10033be18)
 0:12.32  #6 0x000000010ab41c34 std::__1::__function::__func<llvm::parallelFor(unsigned long, unsigned long, llvm::function_ref<void (unsigned long)>)::$_0, std::__1::allocator<llvm::parallelFor(unsigned long, unsigned long, llvm::function_ref<void (unsigned long)>)::$_0>, void ()>::operator()() (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb1dc34)
 0:12.32  #7 0x000000010ab41b60 std::__1::__function::__func<llvm::parallel::TaskGroup::spawn(std::__1::function<void ()>, bool)::$_0, std::__1::allocator<llvm::parallel::TaskGroup::spawn(std::__1::function<void ()>, bool)::$_0>, void ()>::operator()() (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb1db60)
 0:12.32  #8 0x000000010ab411c4 llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::work(llvm::ThreadPoolStrategy, unsigned int) (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb1d1c4)
 0:12.33  #9 0x000000010ab4136c void* std::__1::__thread_proxy[abi:un170006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::ThreadPoolExecutor(llvm::ThreadPoolStrategy)::'lambda'()::operator()() const::'lambda'()>>(void*) (/Users/jakechampion/.mozbuild/clang/lib/libLLVM.dylib+0xb1d36c)
 0:12.33 #10 0x000000019433af94 (/usr/lib/system/libsystem_pthread.dylib+0x180446f94)
 0:12.33 #11 0x0000000194335d34 (/usr/lib/system/libsystem_pthread.dylib+0x180441d34)
```

**Comment by aheejin** [comment]
Even if it does not always fail, we'd appreciate a reproducer. Can you provide one?

**Comment by JakeChampion** [comment]
I've now landed enough of the patches to make the reproduction hopefully simpler to do.

This happens when compiling Mozilla's SpiderMonkey to WASI in debug mode and with Intl included

A full log output can be found at https://treeherder.mozilla.org/logviewer?job_id=468809046&repo=try&lineNumber=5901

Gecko-Dev, which contains SpiderMonkey is a rather large codebase though, the steps to get it building are quite a bit -- https://firefox-source-docs.mozilla.org/setup/macos_build.html

Once that is done you can create a mozconfig file named 'mozconfig-intl-debug' containing:
```
ac_add_options --enable-project=js
ac_add_options --enable-application=js
ac_add_options --target=wasm32-unknown-wasi
ac_add_options --without-system-zlib
ac_add_options --disable-jit
ac_add_options --disable-shared-js
ac_add_options --disable-shared-memory
ac_add_options --disable-tests
ac_add_options --disable-clang-plugin
ac_add_options --enable-jitspew
ac_add_options --enable-optimize=-O3
ac_add_options --enable-js-streams
ac_add_options --enable-portable-baseline-interp
ac_add_options --prefix=${working_dir}/${objdir}/dist
mk_add_options MOZ_OBJDIR=${working_dir}/${objdir}
mk_add_options AUTOCLOBBER=1
ac_add_options --host=aarch64-apple-darwin
ac_add_options --enable-debug
```

And building the project with `MOZCONFIG=./mozconfig-intl-debug ./mach build`

**Comment by aheejin** [comment]
Can you possibly generate a reproducer using `wasm-ld --reproduce`?

**Comment by JakeChampion** [comment]
Thanks, I've generated that file, it's 574M in size, not too sure where I could upload that

**Comment by aheejin** [comment]
You can upload it to a Google drive or Dropbox or something and paste the link here.

**Comment by JakeChampion** [comment]
@aheejin here it is https://mega.nz/file/VlNnRTSK#2kj3qBJFjlgQjl_vOT2MRfZQAG849r1sC9iXvF4qkRs

**Comment by guybedford** [comment]
I'll be picking up from @JakeChampion on this Fastly / BytecodeAlliance error report. If there is any more info that can be provided please just let me know.

**Comment by aheejin** [comment]
Sorry for the long delay. I tried to build the reproducer but I don't seem to have the right WASI environment set..?
```console
$ wasm-ld @response.txt 
wasm-ld: warning: unknown -z value: noexecstack
wasm-ld: warning: unknown -z value: text
wasm-ld: warning: unknown -z value: relro
wasm-ld: warning: unknown -z value: now
wasm-ld: warning: unknown -z value: nocopyreloc
wasm-ld: error: unable to find library -lm
wasm-ld: error: unable to find library -lwasi-emulated-process-clocks
wasm-ld: error: unable to find library -lwasi-emulated-getpid
wasm-ld: error: unable to find library -lc++
wasm-ld: error: unable to find library -lc++abi
wasm-ld: error: unable to find library -lc
```

How am I supposed to set up the libraries?

**Comment by JakeChampion** [comment]
Do you have WASI SDK installed?
https://github.com/WebAssembly/wasi-sdk

**Comment by aheejin** [comment]
Thanks I installed WASI SDK using the prebuilt binarires.

By the way, what's your LLVM version (release version or git commit)?

**Comment by sbc100** [comment]
> Sorry for the long delay. I tried to build the reproducer but I don't seem to have the right WASI environment set..?
> 
> ```
> $ wasm-ld @response.txt 
> wasm-ld: warning: unknown -z value: noexecstack
> wasm-ld: warning: unknown -z value: text
> wasm-ld: warning: unknown -z value: relro
> wasm-ld: warning: unknown -z value: now
> wasm-ld: warning: unknown -z value: nocopyreloc
> wasm-ld: error: unable to find library -lm
> wasm-ld: error: unable to find library -lwasi-emulated-process-clocks
> wasm-ld: error: unable to find library -lwasi-emulated-getpid
> wasm-ld: error: unable to find library -lc++
> wasm-ld: error: unable to find library -lc++abi
> wasm-ld: error: unable to find library -lc
> ```
> 
> How am I supposed to set up the libraries?


Looks like there could be an absolute path in the response file that you would need to update?

**Comment by aheejin** [comment]
@sbc100 Yeah I installed WASI SDK and updated the `-L` path in `response.txt`.

**Comment by aheejin** [comment]
@sbc100 While you're here, do you have any idea what the problem could be?

**Comment by sbc100** [comment]
Looks like a crash in wasm-ld but I can't see much more from the backtrace.   Do you have a backtrace from a LLVM build that includes asserts?

Feel free to assign to me, although I'm not sure when I'll be able to get to it.

**Comment by aheejin** [comment]
What's your LLVM version? Actually I think this bug might have been already fixed in #96134. 

When I tried to reproduce the bug, the tip-of-tree LLVM didn't error out but the last release (LLVM 18) errored out for me. So I bisected this on the LLVM repo and #96134 was the first ~bad~good commit.

If you use LLVM from the source repo, I guess you can simply update the LLVM to the tip of the tree (or any revision equal to or later than #96134). If you use the official LLVM release, LLVM 19 should contain this.

**Comment by aheejin** [comment]
Will close this issue as completed. Feel free to reopen the issue if #96134 does not fix your problem.

**Comment by guybedford** [comment]
I just tried this again today with the latest LLVM 19.1.5 through WASI SDK 25, and am still getting a clang segfault here unfortunately.

Here is the stack:

```
 2:00.22 wasm-ld: warning: function signature mismatch: _ZN7mozilla4intl8Calendar23LegacyIdentifierToBcp47EPKci
 2:00.22 >>> defined as () -> void in ../build/libjs_static.a(Unified_cpp_js_src5.o)
 2:00.22 >>> defined as (i32, i32, i32) -> void in ../build/libjs_static.a(Unified_cpp_intl_components0.o)
 2:00.22 wasm-ld: warning: function signature mismatch: _ZN7mozilla4intl8Collator28KeywordValueToBcp47ExtensionEPKci
 2:00.22 >>> defined as () -> void in ../build/libjs_static.a(Unified_cpp_js_src4.o)
 2:00.22 >>> defined as (i32, i32, i32) -> void in ../build/libjs_static.a(Unified_cpp_intl_components0.o)
 2:00.30 PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
 2:00.56  #0 0x0000000109b3fdf0 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb83df0)
 2:00.57  #1 0x0000000109b4035c SignalHandler(int) (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb8435c)
 2:00.57  #2 0x0000000195fc4184 (/usr/lib/system/libsystem_platform.dylib+0x180484184)
 2:00.57  #3 0x000000010435e784 lld::wasm::InputChunk::relocate(unsigned char*) const (/Users/gbedford/.mozbuild/clang/bin/lld+0x100322784)
 2:00.57  #4 0x000000010435e784 lld::wasm::InputChunk::relocate(unsigned char*) const (/Users/gbedford/.mozbuild/clang/bin/lld+0x100322784)
 2:00.57  #5 0x0000000104377e18 lld::wasm::CustomSection::writeTo(unsigned char*) (/Users/gbedford/.mozbuild/clang/bin/lld+0x10033be18)
 2:00.57  #6 0x0000000109ad9c34 std::__1::__function::__func<llvm::parallelFor(unsigned long, unsigned long, llvm::function_ref<void (unsigned long)>)::$_0, std::__1::allocator<llvm::parallelFor(unsigned long, unsigned long, llvm::function_ref<void (unsigned long)>)::$_0>, void ()>::operator()() (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb1dc34)
 2:00.57  #7 0x0000000109ad9b60 std::__1::__function::__func<llvm::parallel::TaskGroup::spawn(std::__1::function<void ()>, bool)::$_0, std::__1::allocator<llvm::parallel::TaskGroup::spawn(std::__1::function<void ()>, bool)::$_0>, void ()>::operator()() (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb1db60)
 2:00.57  #8 0x0000000109ad91c4 llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::work(llvm::ThreadPoolStrategy, unsigned int) (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb1d1c4)
 2:00.57  #9 0x0000000109ad936c void* std::__1::__thread_proxy[abi:un170006]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, llvm::parallel::detail::(anonymous namespace)::ThreadPoolExecutor::ThreadPoolExecutor(llvm::ThreadPoolStrategy)::'lambda'()::operator()() const::'lambda'()>>(void*) (/Users/gbedford/.mozbuild/clang/lib/libLLVM.dylib+0xb1d36c)
 2:00.57 #10 0x0000000195f8f2e4 (/usr/lib/system/libsystem_pthread.dylib+0x18044f2e4)
 2:00.57 #11 0x0000000195f8a0fc (/usr/lib/system/libsystem_pthread.dylib+0x18044a0fc)
```

I don't have permissions to reopen this issue, please reopen or alternatively I can post this as a new issue.

**Comment by sbc100** [comment]
@guybedford can you try with llvm tip-of-tree, so we know if its something that just need backporting?

**Comment by guybedford** [comment]
@sbc100 thanks for the follow-up. I just ran a full build from trunk and can confirm it is all building and working correctly, so this can be marked as resolved after all.

Really appreciated for the support here.

**Comment by tschneidereit** [comment]
@sbc100 is this something that could perhaps indeed be backported to reduce the time until we can make it available in WASI-SDK?

**Comment by guybedford** [comment]
To clarify - when building from trunk I was able to finally track this down to a pathing issue between the SDKs resulting in an older version than 19 being pulled in. So this is fixed in the current release version as described in the initial diagnosis.

**Comment by tschneidereit** [comment]
Ah, very nice!
